docs/new-feature-tutorial.rst

```diff
@@ -2,84 +2,178 @@
 
-Perhaps your new option controls some other backend rendering: in our case
-we test for this option in the `home` method for adding a variable to the response.
-The functions in this file control the generation of various pages served
-(along with the Django templates).
-Our new feature also shows up in the administration tab (as a checkbox),
-so we need to update the `update_realm` function.
+Frontend changes
+----------------
 
+**JavaScript:** Zulip's JavaScript is located in the directory ``static/js/``. 
+The exact files you may need to change depend on your feature. If you've added a 
+new event that is sent to clients, be sure to add a handler for it to
+``static/js/server_events.js``.
 
-Finally, add tests for your backend changes; at the very least you
-should add a test of your event data flowing through the system in
-``test_events.py``.
+**CSS:** The primary CSS file is ``static/styles/zulip.css``. If your new 
+feature requires UI changes, you may need to add additional CSS to this file.
 
+**Templates:** Zulip uses Handlebars templates which are located in
+``static/templates``. Templates are precompiled as part of the build/deploy
+process.
 
-Frontend changes
-----------------
+**Testing:** There are both blackbox, end-to-end front end tests and unit tests. 
+The blackbox tests are run in a headless browser using Casper.js and are located
+in ``zerver/tests/frontend/tests/``. The unit tests use Node's ``assert`` module
+are located in ``zerver/tests/frontend/node/``. For more information on testing
+see the :doc:`testing documentation <testing>`.
 
-You need to change various things on the front end.  In this case, the relevant files
-are `static/js/server_events.js`, `static/js/admin.js`, `static/styles/zulip.css
-and `static/templates/admin_tab.handlebars`.
+Example Feature
+===============
 
+This example describes the process of adding a new setting to Zulip:
+a flag that restricts inviting new users to admins only (the default behavior
+is that any user can invite other users).
+
+First, update the database and model to store the new setting. Add a 
+new boolean field, ``realm_invite_by_admins_only``, to the Realm model in
+``zerver/models.py``.
+
+Then create a Django migration that adds a new field, ``invite_by_admins_only``,
+to the ``zerver_realm`` table.
+
+In ``zerver/lib/actions.py``, create a new function named 
+``do_set_realm_invite_by_admins_only``. This function will update the database
+and trigger an event to notify clients when this setting changes. ::
+
+  # zerver/lib/actions.py
+
+  def do_set_realm_invite_by_admins_only(realm, invite_by_admins_only):
+    realm.invite_by_admins_only = invite_by_admins_only
+    realm.save(update_fields=['invite_by_admins_only'])
+    event = dict(
+      type="realm",
+      op="update",
+      property='invite_by_admins_only',
+      value=invite_by_admins_only,
+    )
+    send_event(event, active_user_ids(realm))
+    return {}
+
```

Comment:
Maybe note here how we selected the event type, which is that we noticed there's an existing "realm|update" event type supporting multiple properties for similar flags on the Realm model, and so we're just adding this one using that event type with the new property name, which we set to be the name of the field in the database to make things nice and easy to understand.